src/widget.ts

```diff
@@ -104,14 +104,33 @@ export class DeephavenView extends DOMWidgetView {
     }
   };
 
+  //@ts-ignore
+  onDisconnect = (type): void => {
+    log.info(`Kernel ${type}, removing iframe`);
+    this.iframe.remove();
+  }
+
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   render(): any {
     const iframeUrl = this.model.get('iframe_url');
     const width = this.model.get('width');
     const height = this.model.get('height');
+    let removed = false;
 
     window.addEventListener('message', this.handleAuthentication);
 
+    //@ts-ignore
+    const context = this.model.widget_manager._context;
+
+    this.model.on('change:kernel_active', () => this.onDisconnect("exiting"), this)
```

Comment:
I think you should move these listeners to the constructor instead, that should get rid of multiple spawning. See https://github.com/jupyter-widgets/ipywidgets/blob/47058a373d2c2b3acf101677b2745e14b76dd74b/packages/base/src/widget.ts#L992 for an example.

Reply:
@mofojed I'll move it to the constructor, although I was unclear in my original post. When I say multiple listeners, I mean each widget spawns a new listener. Obviously this is expected, the problem is they seem to stick around, even when the widget is gone.